flang/lib/Optimizer/Analysis/AliasAnalysis.cpp

```diff
@@ -60,7 +60,15 @@ void AliasAnalysis::Source::print(llvm::raw_ostream &os) const {
   attributes.Dump(os, EnumToString);
 }
 
-bool AliasAnalysis::Source::isPointerReference(mlir::Type ty) {
+bool AliasAnalysis::isRecordWithPointerComponent(mlir::Type ty) {
+  auto eleTy = fir::dyn_cast_ptrEleTy(ty);
```

Comment:
Not directly related to your patch (since you only moved that code), but I found it while testing the logic below.

This should actually be `fir::unwrapSequenceType(dyn_cast_ptrOrBoxEleTy(ty))` in order to also unwrap fir.box type and arrays.

Otherwise, this will miss cases as the one below where the descriptor address of "p" may alias with the data of "c".

```
module m                                                           
  type t                                                           
     real, pointer :: p                                                                
  end type                                                            
  type(t) :: a(1)                                                           
  type(t) :: b(1)                                                           
contains                                                            
subroutine test(p, c)                                                           
  real, pointer :: p                                                           
  type(t), target :: c(:)                                                              
  p = 42                                                            
  c = b                                                           
  print *, p                                                           
end subroutine                                                            
end module                                                            
                                                            
  use m                                                           
  real, target :: x1 = 1                                                               
  real, target :: x2 = 2                                                               
  a(1)%p => x1                                                           
  b(1)%p => x2                                                           
  call test(a(1)%p, a)                                                                 
end  
```

Reply:
Thanks for pointing that out.  After some of the other issues settle, I'll think about applying that, perhaps in another PR.